message/rpc_protocol_impl.hh

```diff
@@ -180,24 +180,30 @@ auto send_message_cancellable(messaging_service* ms, messaging_verb verb, msg_ad
     auto rpc_client_ptr = ms->get_rpc_client(verb, id);
     auto& rpc_client = *rpc_client_ptr;
 
-    auto c = std::make_unique<seastar::rpc::cancellable>();
-    auto& c_ref = *c;
-    auto sub = as.subscribe([c = std::move(c)] () noexcept {
-        c->cancel();
+    struct state {
+        seastar::rpc::cancellable c;
+        std::exception_ptr ex;
+        optimized_optional<abort_source::subscription> sub;
+    };
+    auto st = make_lw_shared<state>();
+    auto& c_ref = st->c;
+    st->sub = as.subscribe([st] (const std::optional<std::exception_ptr>& opt_ex) noexcept {
+        st->c.cancel();
+        st->ex = opt_ex.value_or(std::make_exception_ptr(abort_requested_exception{}));
     });
-    if (!sub) {
+    if (!st->sub) {
```

Comment:
Doesn't the meaning change? Now it's checking the optimized_optional rather than sub itself.

Reply:
No, it was optimized_optional before too.
That's what subscribe returns.
See https://github.com/scylladb/seastar/blob/502f1422d6e3958237eadcada534fe46a734c8b3/include/seastar/core/abort_source.hh#L134